pallets/auctions/src/lib.rs

```diff
@@ -160,171 +154,297 @@ pub mod pallet {
-			auction.minimal_bid = value.checked_add(&minimal_bid_step).ok_or(Error::<T>::BidOverflow)?;
-			// Avoid auction sniping
-			let time_left = auction
-				.end
-				.checked_sub(&block_number)
-				.ok_or(Error::<T>::TimeUnderflow)?;
-			if time_left < BID_ADD_BLOCKS.into() {
-				auction.end = block_number + BID_ADD_BLOCKS.into();
+	fn validate_bid(
+		bidder: &<T>::AccountId,
+		general_auction_data: &GeneralAuctionData<T>,
+		value: BalanceOf<T>,
+	) -> DispatchResult {
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		ensure!(bidder != &general_auction_data.owner, Error::<T>::BidOnOwnAuction);
+		ensure!(block_number > general_auction_data.start, Error::<T>::AuctionNotStarted);
+		ensure!(
+			block_number < general_auction_data.end,
+			Error::<T>::AuctionEndTimeReached
+		);
+		ensure!(value >= general_auction_data.next_bid_min, Error::<T>::InvalidBidPrice);
+
+		if let Some(ref current_bid) = general_auction_data.last_bid {
+			ensure!(value > current_bid.1, Error::<T>::InvalidBidPrice);
+		} else {
+			ensure!(!value.is_zero(), Error::<T>::InvalidBidPrice);
+		}
+
+		Ok(())
+	}
+
+	fn validate_close(general_auction_data: &GeneralAuctionData<T>) -> DispatchResult {
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		ensure!(!general_auction_data.closed, Error::<T>::AuctionClosed);
+		ensure!(block_number >= general_auction_data.end, Error::<T>::AuctionEndTimeNotReached);
+
+		Ok(())
+	}
+}
+
+impl<T: Config> NftAuction<T::AccountId, T::AuctionId, BalanceOf<T>, Auction<T>> for EnglishAuction<T> {
+	fn bid(&mut self, bidder: T::AccountId, value: BalanceOf<T>) -> DispatchResult {
+		// Lock / Unlock funds
+		if let Some(ref current_bid) = self.general_data.last_bid {
+			<T as crate::Config>::Currency::remove_lock(AUCTION_LOCK_ID, &current_bid.0);
+		}
+		<T as crate::Config>::Currency::set_lock(AUCTION_LOCK_ID, &bidder, value, WithdrawReasons::all());
+
+		self.general_data.last_bid = Some((bidder, value));
+		// Set next minimal bid
+		let bid_step = Permill::from_percent(<T as crate::Config>::BidStepPerc::get()).mul_floor(value);
+		self.general_data.next_bid_min = value.checked_add(&bid_step).ok_or(Error::<T>::BidOverflow)?;
+
+		// Avoid auction sniping
+		let block_number = <frame_system::Pallet<T>>::block_number();
+		let time_left = self
+			.general_data
+			.end
+			.checked_sub(&block_number)
+			.ok_or(Error::<T>::TimeUnderflow)?;
+		if time_left < <T as crate::Config>::BidAddBlocks::get().into() {
+			self.general_data.end = block_number + <T as crate::Config>::BidAddBlocks::get().into();
+		}
+
+		Ok(())
+	}
+
+	fn close(&mut self) -> DispatchResult {
+		pallet_uniques::Pallet::<T>::thaw(
+			RawOrigin::Signed(self.general_data.owner.clone()).into(),
+			self.general_data.token.0,
+			self.general_data.token.1,
+		)?;
+
+		// there is a bid so let's determine a winner and transfer tokens
+		if let Some(ref winner) = self.general_data.last_bid {
```

Comment:
nitpick : i find this way more readable.

```suggestion
		if let Some(winner) = &self.general_data.last_bid {
```